### ISSUE TITLE
Weekly snapshots not being retained

### ISSUE THREAD

**Issue by daveTheOldCoder** [bit-team/backintime#1094]
I'm using the default values for Smart Remove. Hourly and daily snapshots are kept or deleted as expected, but no weekly snapshots are kept.

This seems to be the relevant code in /usr/share/backintime/common/snapshots.py:
```
        #keep one per week for the last keep_one_per_week weeks
        if keep_one_per_week > 0:
            d = now - datetime.timedelta( days = now.weekday() + 1 )
            for i in range( 0, keep_one_per_week ):
                keep_snapshots = self._smart_remove_keep_first_( snapshots, keep_snapshots, d, d + datetime.timedelta(days=8) )
                d = d - datetime.timedelta(days=7)
```
I'm trying to figure it out, but I thought I'd ask first if there's a known issue.

**Comment by daveTheOldCoder** [comment]
I found a potential fix for this problem.

The function "_smart_remove_keep_first_" keeps the newest snapshot in the specified date/time range, since the snapshots are sorted in descending order of date/time. I added a function "_smart_remove_keep_last_" that reverses the order of the snapshots and thus keeps the oldest snapshot.

In the code in the previous post, I replaced the call to "_smart_remove_keep_first_" with a call to "_smart_remove_keep_last_". So far it seems to be working. But it needs further testing and analysis.

If this is actually a bug, I don't understand why it hasn't been reported before. Can someone else review the code/algorithm and provide an opinion?

**Comment by buhtz** [comment]
@daveTheOldCoder There is a new forming maintaining team and we do review all issues. Is this problem still relevant for you or did you find a solution?

Tag: Feedback, Bug

**Comment by daveTheOldCoder** [comment]
I'm still using the fix I described above, and it works properly. Please review my posts above. Do you agree with my analysis of the problem?

I can provide the exact changes I made, if that's helpful.

And thanks for taking on the job of maintaining this product. :)

**Comment by daveTheOldCoder** [comment]
Also: I'm using BackInTime version 1.2.1 (the latest version available with the package manager on Pop!_OS 20.04), with my changes added. I'll look at the latest files in the github repository and see if there are relevant changes. Since this issue had no replies prior to yours, I assumed that the issue had not been fixed.

**Comment by daveTheOldCoder** [comment]
I just looked at BackInTime version 1.3.2, and the relevant code in common/snapshots.py is unchanged from version 1.2.1.

**Comment by daveTheOldCoder** [comment]
The fix I made was a one-liner.

In def smartRemoveKeepFirst, change:
`for sid in snapshots:`
to:
`for sid in reversed(snapshots):`

But it would probably be "cleaner" to rename smartRemoveKeepFirst to smartRemoveKeepLast, and then change the calls to it.

**Comment by buhtz** [comment]
Dear Dave,
sorry for let you wait on this and thanks for your effort and your suggested fix.

The problem is not only resources but the test coverage. The smart remove behavior is not covered by unit tests. Most of the tests in BIT are system or integration tests. And I assume the related code is not well isolated to create a real unit test for it.
Because I am not the founder of BIT and not deep enough into all details I am worried to modify this part of the code without test coverage. Introducing bugs into this behavior can be a huge problem and effect a lot of people.

Would you like to jump in? We take you by hand and mentoring you if you need it.
So the **first** job would be to analyze the testability of the smart-remove code. **Second** would be to carefully refactor and modify that code, hopefully without modifying its behavior, to make it testable. **Third** job would be to create real and isolated unit tests to fully cover the current behavior. **Forth** job then is to introduce your fix, running tests on it etc pp.

**EDIT**:
This is how the relevant code section (in `common/snapshots.py::Snapshots.smartRemoveList`) looks today:
https://github.com/bit-team/backintime/blob/e818aeb2b009e2be052685c70dae202c88a00aee/common/snapshots.py#L1674-L1684

I couldn't resist and took a closer look about testabillity. As I thought it is nearly untestable without heavy mocking and hacking. Beside the following points I assume there will come up even more when looking close and longer. Based on my short observation the following points need refactoring to make the behavior testable:
 - I see no reason for existence of an instance of the class `Snapshots`. All its methods should live as module functions. That class holds the logic of the application layer related to snapshot handling.
 - The method `smartRemoveKeepFirst()` need to be decoupled from the `class SID`. btw: An instance of `SID` represent a specific snapshot. The same is for `smartRemoveKeepAll()`.
 - `smartRemoveList()` and `smartRemove()` are definitely to long. Separate into several functions or methods. Maybe here is an opportunity to create a class dedicated to this behavior. But all this remove logic could also live in its own modul.
 - `SID` and `Snapshots` shouldn't live in the same module because they serve different layers.
 - And of course, one of BITs main problems in the code structure, decouple `Config` and its instance from the whole module. The final golden goal is to remove the `import config` from there.

It is much and I just added this points for documentation. I don't think that all aspects need to be full filled to make the target behavior testable. I would say that "decouple from `SID`" would be a good start. Implementing this will take multiple PRs.

**Comment by daveTheOldCoder** [comment]
I still don't understand why no one else has reported this problem. For me, it's a critical problem. I wouldn't be able to use BackInTime without my fix.

**Comment by buhtz** [comment]
Good question. Can you send us a screenshot of your smart remove tab please.

**Comment by daveTheOldCoder** [comment]
![screenshot](https://github.com/bit-team/backintime/assets/55119684/597c03f4-98a9-4014-954a-62a62928e8b2)

**Comment by buhtz** [comment]
Your screenshot looks OK for me.

I checked my own setup. But my machine do not run every day. So I can not be sure.
![grafik](https://github.com/bit-team/backintime/assets/11861496/def5d72a-eb8a-4f84-b158-d45bdc944554)

**Comment by buhtz** [comment]
> This can be fixed with a one-line change in common/snapshots.py:
> 
> In `def smartRemoveKeepFirst`, change: `for sid in snapshots:` to: `for sid in reversed(snapshots):`
> 
> But it would probably be "cleaner" to rename `smartRemoveKeepFirst` to `smartRemoveKeepLast` or simply `smartRemove`, and then change the calls to it.

I don't get how this solve your problem. Why does it matter if the first or the last snapshot in a week-range is kept? To my understanding this does not explain why you lose the weekly snapshots.

Maybe you can explain it better with same example dates?

**Comment by daveTheOldCoder** [comment]
It's been almost four years since I submitted the issue. I'll try to find my notes.

**Comment by emtiu** [comment]
A way to efficiently test this issue might be to manipulate the date/time of a VM to simulate the passing of weeks and months quickly.

**Comment by emtiu** [comment]
I have created a testbed with the following shell code in a VM:
```for N in {1..60}; do sudo date -s '+7 days'; echo date > backupfile; backintime backup; done```

In extreme cases, it appears to work as expected: Keeping _"one snapshot per week for the last **52** weeks"_ really does leave 1 years' worth of weekly snapshots. However, I'm not so sure about shorter timespans. Calculating dates is hard!

My current guess is that we may be dealing with an off-by-1 problem here, where a setting of _"one snapshot per week for the last **n** weeks"_ only actually keeps one for the last _**n–1** weeks_. This would be especially significant for the default configuration of:
* _"one snapshot per day for the last **7** days"_
* _"one snapshot per week for the last **4** weeks"_
* _"one snapshot per month for the last **24** months"_

**Comment by buhtz** [comment]
I once had a tool called [`faketime`](https://tracker.debian.org/pkg/faketime) I used for things like this. Worked well.

**Comment by emtiu** [comment]
I think I'm very close to cracking this case. In this function:
https://github.com/bit-team/backintime/blob/22f468c32732be6c821686e0c504b4dfb920487a/common/snapshots.py#L1686-L1696
… `smartRemoveKeepFirst` is called with a `max_date` of `d + datetime.timedelta(days=8)` — not `days=7`!

Therefore, in a routine that says "keep only the _youngest_ snapshot of these _8 days_ to consider", it will also delete the youngest snapshot of the _previous_ week — but only if it's caught by the **_last_** in a series of "keep one per week" calls.

In effect, of _"one snapshot per week for the last **n** weeks"_, the **n**-th week has its youngest snapshot thrown away, even though it shouldn't.

I'll need to investigate if changing the call of `smartRemoveKeepFirst` to a `max_date` of `d + datetime.timedelta(days=7)` fixes the problem. There must have been _some_ reason to set an `8` there.

P.S.: The official soundtrack for this bug is [**Only For The Week**](https://www.youtube.com/watch?v=qnLunQEcMn0).

**Comment by buhtz** [comment]
There are some unit tests in `test_snapshot.py` about that smart remove thing.
I touched some once long ago but then gave up. A full refactoring of that tests would be the best. But a first step would be to see if this use case is covered by that tests and add such a test if not.

**Comment by daveTheOldCoder** [comment]
It occurred to me that what makes the "smart remove" behavior confusing is that the snapshots are named using only the timestamp. The names don't contain "hourly", "daily", "weekly", etc.

I wonder if adding an index file that identifies daily, weekly, monthly and yearly snapshots would help.

Or maybe I'm wrong. :)

**Comment by emtiu** [comment]
> There are some unit tests in `test_snapshot.py` about that smart remove thing. I touched some once long ago but then gave up. A full refactoring of that tests would be the best. But a first step would be to see if this use case is covered by that tests and add such a test if not.

I'm confident that I can fix this bug, and to make sure I don't introduce any new ones, I'm doing some manual testing in a VM.

I'm totally unfamiliar with automatic testing, though. I'll take a look, but I'd rather not put this bugfix on the "waiting list" until I've cracked the automatic testing of Smart Remove, because that would probably take a very long time ;)

**Comment by emtiu** [comment]
> It occurred to me that what makes the "smart remove" behavior confusing is that the snapshots are named using only the timestamp. The names don't contain "hourly", "daily", "weekly", etc.
> 
> I wonder if adding an index file that identifies daily, weekly, monthly and yearly snapshots would help.
> 
> Or maybe I'm wrong. :)

I see what you mean, but BiT doesn't work like that. Snapshots are not created as "weekly" or "monthly" snapshots. In fact, they are only characterized by their date+time of saving (plus a random three-digit number, which is called the "tag"). Together, this makes a "Snapshot ID" (SID for short): `20241117-115358-562`

The only time when a snapshot might be characterized as "weekly" or "monthly" is when the Smart Remove routine determines: _"this snapshot should be kept, because it's the only one from that week/month, and it should be kept according to the removal rules"_.

Therefore, any single snapshot can start out as an "hourly" snapshot, then become a "weekly" one, and maybe even a "monthly" one later.

**Comment by daveTheOldCoder** [comment]
>I see what you mean, but BiT doesn't work like that.

I know that. To figure out the [workaround](https://github.com/bit-team/backintime/issues/1094#issuecomment-636493410) I posted above, I added logging code in the smart-remove processing to log the "kept" and "removed" snapshots, so I could see what was happening.

**Comment by emtiu** [comment]
> I know that. To figure out the [workaround](https://github.com/bit-team/backintime/issues/1094#issuecomment-636493410) I posted above, I added logging code in the smart-remove processing to log the "kept" and "removed" snapshots, so I could see what was happening.

I see, and I've also understood why your workaround is successful. It's hard to put into words ;)

What happens is that the code considers snapshots from _8 days_ (instead of 7) when determining which to keep for a specific week. But in a one-snapshot-per-week environment, 8 days will likely contain _two_ snapshots. Since `smartRemoveKeepFirst` keeps only the _newer_ one, the older one is lost.

This has no consequences as long as the previous week is also considered before deletion. But it fails for the _last_ of the weeks to be considered (resulting in the _**n–1**_ situation I described above).

Your workaround reverses the list of snapshots, so in the "8-day-week" that is the last to be considered, the _older_ of the two snapshots is kept, and your problem disappears.

However, it is my understanding that the problem disappears completely when `days=8` is changed into `days=7` :)

**Comment by daveTheOldCoder** [comment]
After verifying that the bug was still present in version 1.4.3 (the latest version available for Ubuntu 24.04), I'm testing your fix (days=8 -> days=7) by making that one-line change in that version.

It will take a few more weeks. I want to verify that the monthly snapshots are retained too.

**Comment by buhtz** [comment]
> It will take a few more weeks. I want to verify that the monthly snapshots are retained too.

Cool! Real time testing is the best. :smile:

**Comment by daveTheOldCoder** [comment]
Ideally, I'd like to write a simulator for the code in snapshots.py that could do the testing in seconds.

But I'm too lazy and senile to accomplish that now. :)

**Comment by buhtz** [comment]
Hello Guys,
After working on PR #1944 (some unit tests for this topic) and investigating emtiu's PR #1819 I am not sure what the problem is.
Dave, your initial post did not describe the problem in details. What is the problem? What is the actual behavior and what is the expected behavior?

I was not able to produce a problem in my testings.

The actual behavior I have observed in the context of "keep one snapshot per week for the last N weeks".
- The current week is ignored. The calculation starts with the previous week and goes back in time from there.
- It keeps always the last/newest backup in a week. If there is one backup for each day in the week it will keep always the backup from Sunday.

**Comment by daveTheOldCoder** [comment]
It's been 4-1/2 years since I posted the original problem, and I don't remember the exact details.

The fix I posted above works for me.

I'm going to use the released version of BackInTime with my fix applied.

**Comment by buhtz** [comment]
Thank you for the reply. 
Will wait for response from Michael. He might bring light into the topic. 

Would be nice to have example Dates to reproduce.

**Comment by daveTheOldCoder** [comment]
Ideally, the relevant code should be simple and clear enough that its correctness, or lack of correctness, is obvious.

If code is too complex to understand easily, then it's too complex.

But I'm not volunteering to rewrite it. ;)

**Comment by buhtz** [comment]
Hello Dave,
Please allow me to add another question to you.

Out of your guts, does the original "keep one per N weeks" settings include the current running and not complete week or not? From your perspective as a user what would you expect?

> The actual behavior I have observed in the context of "keep one snapshot per week for the last N weeks".
> 
> * The current week is ignored. The calculation starts with the previous week and goes back in time from there.
> * It keeps always the last/newest backup in a week. If there is one backup for each day in the week it will keep always the backup from Sunday.

It seems my assumption was not correct. The algorithm try to take the current week also into account. But that does not work in all situations. I am on it...

**Comment by daveTheOldCoder** [comment]
Since Oct 10, I've been using BackInTime 1.4.3 with the one-line change in this PR manually applied:
https://github.com/bit-team/backintime/pull/1819/commits/a207fc28e68beef41101c26ac75f0a8b4d42c5b3

BackInTime was run hourly, with the same Auto-remove settings as in the [screenshot](https://github.com/bit-team/backintime/issues/1094#issuecomment-2010299705) I posted above on Mar 20.

Monthly and hourly snapshots are being retained, but I think that the number of weekly backups retained is one less than it should be. For that reason, I recently changed the weekly setting to: "Keep one snapshot per week for the last 5 weeks".

If there is ambiguity in the meaning of the setting, I think that a user would prefer to retain one more, rather than one less, backup. You can remove an extra backup; you can't restore a deleted one.

Here is the current list of snapshots:
20240930-230001-367
20241031-230002-477
20241102-220001-447
20241109-230002-184
20241116-230001-201
20241120-230002-530
20241121-230002-290
20241122-220001-856
20241123-230001-801
20241124-230001-181
(+ hourly snapshots for 20241125 and 20241126)

The same issue may apply to the daily snapshots. It seems like the number retained is smaller than it should be.

When I was using BackInTime 1.2.1, with the [patch](https://github.com/bit-team/backintime/issues/1094#issuecomment-1257264064) I posted above on Sep 25, 2022, I didn't observe any problems.

**Comment by buhtz** [comment]
Thank you for providing your data and setup.

I agree that there are "problems" with that algorithm. According to my current understanding of it you do have 5 weekly backups in your list. Some of the backups in your list do hit multiple of of the keep-rules.

- monthly:
  - 30 Sept.
  - 31 Oct.
- weekly:
  - 2 Nov
  - 9
  - 16
  - 23
  - 26 (vanilla BIT) or 25 (in your patched BIT because of `reversed()`)
 - daily:
   - 20 to 26 Nov

You see there are 5 weekly backups. The 25/26 is hit twice by the weekly- and the daily-rule.

Just for my own documentation this might illustrate the situation:
![Gescanntes Dokument](https://github.com/user-attachments/assets/dedb557a-3a42-4c91-af5e-4ea5423419e7)

**Comment by daveTheOldCoder** [comment]
>Out of your guts, does the original "keep one per N weeks" settings include the current running and not complete week or not? From your perspective as a user what would you expect?

On reflection, I consider "keep one for N weeks" to mean: 
Last week
Two weeks ago
Three weeks ago
...
N weeks ago

**Comment by buhtz** [comment]
Is 18. to 24. the "Last week"?

**Comment by daveTheOldCoder** [comment]
In my opinion, "last week's backup" is the snapshot taken approximately seven days ago. It's hard to define precisely.

It's like art. It's hard to define, but you know when you like it.

**Comment by buhtz** [comment]
I see, you define "a week" as a timedelta and not as a calendar element starting on Monday.

**Comment by daveTheOldCoder** [comment]
In this context, I suppose so. I haven't given the matter much thought.

**Comment by buhtz** [comment]
Might not be directly relevant to this issue. But I investigated the current auto-/smart-remove behavior and documented it. I found inconsistent behavior and minor bugs. There is a lot potential to improve and refactoring.

Before starting with this I would like to request your opinions and suggestions about the next steps. Please see the issue #1945 for details and a mockup for the new auto-remove dialog-tab.

Feedback and suggestions welcome. We appreciate your feedback.

**Comment by buhtz** [comment]
Feel free to re-open if PR #2021 not fixing the problem.
Thank you for reporting and participating.

**Comment by daveTheOldCoder** [comment]
Okay, thanks.

I still don't understand why the original problem seemed to disappear, with no relevant changes to the code.

Maybe I'm insane.

**Comment by buhtz** [comment]
There are changes in two lines

https://github.com/buhtz/backintime/commit/1ff3a64f73b1d79149121cf61a2634a8eec1527e#diff-c9ecd4097accca29159e1baf73ee2a101a03d785aa540551a8902aab6abbe07b

Don't worry. The original code is very hard. And I plan to refactor it.
